typing/uniqueness_analysis.ml

```diff
@@ -0,0 +1,1507 @@
+      (Ienv.Extension.conjunct ext0 ext1, UF.seq uf0 uf1)
+  | Tpat_constant _ ->
+      (Ienv.Extension.empty, mark_implicit_borrow_memory_address Read value)
+  | Tpat_tuple pats ->
+      pat_proj
+        ~handle_tuple:(fun values ->
+          (* We matched a tuple against a tuple parent and descend into each
+             case *)
+          (* no borrowing needed - we own the tuple! *)
+          let exts, ufs =
+            List.split
+              (List.map2
+                 (fun pat (value, _) -> pattern_match pat (Match_single value))
+                 pats values)
+          in
+          (Ienv.Extension.conjuncts exts, UF.seqs ufs))
+        ~extract_pat:Fun.id
+        ~mk_proj:(fun i _ -> Usage_tree.Projection.Tuple_field i)
+        value pats
+  | Tpat_construct (lbl, _, pats, _) ->
+      pat_proj ~extract_pat:Fun.id
+        ~mk_proj:(fun i _ ->
+          Usage_tree.Projection.Construct_field (Longident.last lbl.txt, i))
+        value pats
+  | Tpat_variant (lbl, mpat, _) ->
+      let uf = mark_implicit_borrow_memory_address Read value in
+      let t =
+        match value with
+        | Match_single { paths; _ } ->
+            {
+              paths =
+                UF.Path.child_of_many paths
+                  (Usage_tree.Projection.Variant_field lbl);
+              loc = pat.pat_loc;
+            }
+        | Match_tuple _ ->
+            (* matching a tuple against variant can't pass type checking *)
+            assert false
+      in
+      let ext, uf' =
+        match mpat with
+        | Some pat' -> pattern_match pat' (Match_single t)
+        | None -> (Ienv.Extension.empty, UF.unused)
+      in
+      (ext, UF.seq uf uf')
+  | Tpat_record (pats, _) ->
+      pat_proj
+        ~extract_pat:(fun (_, _, pat) -> pat)
+        ~mk_proj:(fun _ (_, l, _) ->
+          Usage_tree.Projection.Record_field l.lbl_name)
+        value pats
+  | Tpat_array (_, pats) -> (
+      match value with
+      | Match_tuple _ -> assert false
+      | Match_single { paths; loc; _ } ->
+          let uf = mark_implicit_borrow_memory_address_paths Read paths loc in
+          let exts, ufs =
+            List.split
+              (List.map
+                 (fun pat ->
+                   let value =
+                     Match_single
+                       { paths = [ UF.Path.fresh_root "array_field" ]; loc }
+                   in
+                   pattern_match pat value)
+                 pats)
+          in
+          (Ienv.Extension.conjuncts exts, UF.seqs (uf :: ufs)))
+  | Tpat_lazy pat' -> pattern_match pat' value
+  | Tpat_or (a, b, _) ->
+      let ext0, uf0 = pattern_match a value in
+      let ext1, uf1 = pattern_match b value in
+      (* note that we use Ienv.par *)
+      (Ienv.Extension.disjunct ext0 ext1, UF.seq uf0 uf1)
+
+and pat_proj :
+      'a.
+      ?handle_tuple:_ ->
+      extract_pat:('a -> _) ->
+      mk_proj:(_ -> 'a -> _) ->
+      _ ->
+      'a list ->
+      _ =
```

Comment:
I think `pat_proj` makes things harder to understand. I think there are three changes that would simplify things and allow us to remove it:
1.  Rather than trying to get all these projection cases to behave differently on `Match_single` and `Match_tuple`, I think it would be better to have them all of them -- other than `Ppat_tuple` -- start by calling the function which turns a `value_to_match` into a `single_value_to_match`. In these cases, the type-checker should prevent us seeing a `Match_tuple` case, but if we somehow did, converting it to a single value and then projecting from that is the only thing that would make sense.
2. Lift all the projections to work on `single_value`s. That would allow lines 952-954 to get replace by a single function call in each instance.
3. Use a `fold_left`, applying `conjunct` and `seq` as we go rather than doing a `map`, a `split` and a list-based operation.

After these changes, I think you will find that the code for each case is simple enough to not need to share an auxiliary function.

Reply:
I believe this is considered resolved.